# CUPP Conserved Unique Peptide Patterns (CUPP-Predict)

v: 3.2.1

### Description

Conserved Unique Peptide Patterns (CUPP) is a new approach for sequence analysis employing conserved peptide patterns for determination of similarities between proteins.

This CUPP command line application represents the implementation of the CUPP predication algorithm in RUST, offering increased annotation speed with a significant reduction in memory overhead.

The CUPP prediction tool is also available as a web server at https://cupp.info/.

The cupp-predict binary file along with the associated CUPP library and meta files, is able to predict putative Carbohydrate-Active enZymes (CAZymes). It outputs the predicted CAZy family, subfamily and EC number of the query sequence.

Precompiled binaries for cupp-predict area available for:

- Windows, x86_64-pc-windows-gnu
- Linux, x86_64-unknown-linux-gnu
- macOS, x86_64-apple-darwin

## Running cupp-rust

### Requirements

- cupp-predict binary for your platform
- CUPP library file with the .lib extension <8x2_90_v1.1.0_transferred_CUPPlibrary.lib>
- CUPP meta file with the .meta extension <8x2_90_v1.1.0_transferred_CUPPlibrary.meta>
- Your query FASTA file containing either protein and nucleotide sequences. CUPP can automatically find ORF in genomic sequences based on the standard genetic code (NCBI transl_table=1) with the standard initiation codon AUG.

### Basic usage

```
cupp-predict -q <PATH_TO_QUERY_FILE> -l <PATH_TO_CUPP_LIB>
```

### Advanced usage

```
USAGE:
    cupp-predict [FLAGS] [OPTIONS] --query <QUERY_FILE> --library <LIB_LOCATION>

FLAGS:
    -x, --experimental    Include annotation by cupp Groups that have no CAZy members
    -h, --help            Prints help information
    -g, --genomic         Treat the query as DNA and locate ORFS and predict them individually
    -k, --keepfasta       Write all query to output fasta, including those without annotations
    -f, --filteroff       Do not do domain filtering, but keep all domains
    -V, --version         Prints version information
    -v, --verbose         Verbose output level

OPTIONS:
        --cuppconfidence <FLOAT>         Threshold confidence level for a cupp Group and EC annotation to be valid
                                         [default: 5.0]
    -c, --cwd <CWD>                      Change working directory for output and temp files [default: ]
    -r, --cratio <FLOAT>                 Number of positions that must be covered within a domain for it to be valid
                                         [default: 0.5]
    -o, --overlap <FLOAT>                How much can a domain overlap with an already assigned domain [default: 0.4]
    -e, --eratio <FLOAT>                 Minimum percentage of frequency of cupp group for hit [default: 0.01]
    -q, --query <QUERY_FILE>             Select a fasta file for cupp prediction REQUIRED
    -l, --library <LIB_LOCATION>         Specify full path to cupp Library or meta file REQUIRED
    -d, --domainmin <INT>                Minimum number of positions covered to be a domain [default: 16]
    -m, --orfminlength <INT>             Minimum number of bases in ORFs. If 0, the sequence will be treated as one ORF
                                         in frame and do not preform ORF searching [default: 60]
    -n, --nproc <INT>                    Number of processors [default: 1]
        --output <output>                Sets the available output modes: BOTH (both the tab and fasta file), TABS (tabs
                                         file only), FASTA (fasta file only) [default: BOTH]  [possible values: BOTH,
                                         TABS, FASTA]
    -p, --posthreshold <FLOAT>           Frequency for a position to be included in the domain [default: 0.2]
        --subfamilyconfidence <FLOAT>    Threshold confidence level for a Subfamily annotation to be valid [default:
                                         3.0]
```

Protein sequences example

```
>EAA29705.1 hypothetical protein [Neurospora crassa]
MVPSLQTVGHILVTALAIIPFAFADLTTTIDPSSNWGTWEGWGTSLAWWAATFGTRDDLA
DIFFTTQTTNLNGVSLPGLGFNIARYNAGASSWNSINGTSMVVSPKMIPSRQIEGFWIDW
ASTSPSSASWNWSADSAQRAMLSKARSRGANILELFSNSPMWWMCNNHNPSGSDDGSSDN
LQSWNYQSHAVYMATVAKYFADHWSTTFTSVDPFNEPSANWWNGKTGTQEGCHFDVSTQA
TVLNYLRAELDSRGLSSTVISASDESYYDQAAKTLQNLGSTALGKIARVNVHGYQYGSGR
RDSVRSLATQYSKKLWQSEYGEGDATGRDLASNLLLDLRWLQPTAWVYWQALDGGGWGLV
DADNDKKTIGQPTQKYFVLAQFARHVKPGMRILDGGSDYATAAYDAATKKLVIVAVNWGS
AQYLNFDLSRFKTAGVNGAVVKRWATQIGASTGKRYVAASDTYIRGTKFWSYFEKEMVQT
FEVEGVVL

>BAO73305.1 LcGUS30 [Lactobacillus brevis subsp. coagulans]
MKKFQHEFWTATSGDLTQRRTPLATPDYQSTSAAAVKIIIDPADRHQPWLGGGAAITDSAAYLLWSVMSAEQRRALLTELFDPDQGGFSSVRVPLGSCDFQSQDFYTYDDVPYGEHDQKLEQFSIGTGQPGAPDATKDLKHIVPVLQEILAINPAVKVIASPWSAPAWMKNTGHLTHGGHLRFGEFTGNGYTEENRFEYIYAQYFIRYIEAYQKLGIPIYGLTIQNEPSNAAHWPAMIWTVPQLADFGYRYLRPALNHSFPDTKLYLLDDSFHALTKPITAEVTPEQAAAFDGLAVHTYSGPYDNLYHANRAYPNWSTIMTERRCMMTDTPEEAAHIMFGIIGNWLVHNGLSMITLWNLALDERGLPNAADSTGREGVVTIDHTTGKVQRNLEYFMLRNFGQDVSVGATVIGSTNYTRDGYTGGLGSVAFLGTAGDIAAHLYNPTAQPIQAAVTINGNGANWQLVTVPPYGTVTLHKSDAPLNTTNVPVDDEFPLNPTPANHSDVAPGQH

>BAE55732.1 unnamed protein product [Aspergillus oryzae RIB40]
MSNSNPPLITHIYTADPSAHVFNNKVYIYPSHDRETDIQFNDNGDQYDMVDYHVFSMDSISGPVTDHGVVLTLSDVPWASKQLWAPDAATKDGKYYLYLPARDHDGIFRIGVAVAEKPEGPFTAQKSFIAGSYSIDPASFVDDDGKAYLYFGGIWGGQLQCWSKDAGAAGEWVFDASKTGPQEPSGENELALHPRVAQLTDDMLGFETPVQELVILDQDGVSPLKADDHERRFFEASWMHKYNGTYYFSYSTGDTHYIVYATGESPFGPFRYRGRILEPVLGWTTHHSIVEFGGRWWIFYHDCELSKGVSHLRSVKVREIGYDDQGDIYLVD
>AAP64786.1 Sequence 2 from patent US 6555335
MKSSISVVLALLGHSAAWSYATKSQYRANIKINARQTYQTMIGGGCSGAFGIACQQFGSSGLSPENQQKVTQILFDENIGGLSIVRNDIGSSPGTTILPTCPATPQDKFDYVWDGSDNCQFNLTKTALKYNPNLYVYADAWSAPGCMKTVGTENLGGQICGVRGTDCKHDWRQAYADYLVQYVRFYKEEGIDISLLGAWNEPDFNPFTYESMLSDGYQAKDFLEVLYPTLKKAFPKVDVSCCDATGARQERNILYELQQAGGERYFDIATWHNYQSNPERPFNAGGKPNIQTEWADGTGPWNSTWDYSGQLAEGLQWALYMHNAFVNSDTSGYTHWWCAQNTNGDNALIRLDRDSYEVSARLWAFAQYFRFARPGSVRIGATSDVENVYVTAYVNKNGTVAIPVINAAHFPYDLTIDLEGIKKRKLSEYLTDNSHNVTLQSRYKVSGSSLKVTVEPRAMKTFWLE

>AAU40201.1 arabinan endo-1,5-alpha-L-arabinosidase AbnA [Bacillus licheniformis DSM 13 = ATCC 14580]
MLKTSKFERRTTVAKTIISGFILFFLLIFSAAEPTSAAFWDTKGDNIIHDPSMIKEGNTWYTFGTGIGNGIRVIKSTDGKTWSAAPSIFATPLSWWKKYVPNHEENQWAPDISYYNGRYWLYYAVSAFGSNTSAIGLASTDRISSGNWRDDGLVIRTTTSNNYNAIDPELVIDKEGNPWLSFGSFWSGIKLTKLDKKTMKPTGKVYSIASRPNNGGAVEAPSITYKNGYYYLFVSFDKCCQGVNSTYKIAYGRSTKITGPYYDKSGKKMTEGGGTILDSGNDQWKGPGGADIVNGNIIVRHGYDALDNGAPKLLINDLNWDSKGWPKY
```

Nucleotide sequences example

```
>XM_001391539.2 Aspergillus niger CBS 513.88 endo-arabinase, mRNA
ATGAAAGGCCTCCCGTCACTGCTGAACAGCAGCAGCACCAAAAAATGGACACAATCCACC
CCATCCACAGGCGAGGGTGAGAATGAGAGTGGTAGTACTATATCACCCACCAAGCCGCCC
GAGCCATCCACACCCCCAGGAGGCTCAAACACCTCCATGATCTGGACGCGGAAGAAGATA
ATATGGCTCATCACCCTGATAGCTCTAAGCGTAGCCATAATCGTGGTGGTGATAACCGTC
CCAGTAGTTCTCCTGCTGGACGATAAACACAACGACGATCCCAGCTACTACAACAGCGCC
AACCGTCCAGTCCGCGTCGTGCATGACTTCCCCGACCCGGGTCTAATCCAGGTGAACAGC
ACGTGGTATGCATACGCTACAGTCGCCACACCGGACAATCCTGACGTCCCTCACGTCCCG
GTATCGACATCGCGGAACTTTAGCAGCTGGACCTGGTTGCAGGGATACGATGTTATGCCT
GCAATTAGCAGTTGGGAAACGAATATGAACCAGTACGCTCCGGACGTGATTCAACGCAAA
GACGGCCACTTCGTCCTCTACTACTCAGGCGAACTAAAAGACTGGCTCCGCCACCACTGC
GTCGGCGCCGCTGTCTCCAACGGCACAAGTCCCCTAGGACCATACATCCCCCACAACACC
ACCCTCGCCTGTCCCCGCGACCACGGCGGTGCCATCGACCCCGCCCCCTTCAGAGACGTC
AACGGGACCCTCTACGTCGTGTACAAAGTCGACGGCAACAGCATCGGCCACGGCGGAGGC
TGCAACAACGGCAAGAAACCCATCGTCTCCACGCCCATCATGCTCCAACAACTCAAAGAC
GATGGAGTCACCCCCGTCGGTGATCCCGTTGAAATTCTAACCAACGAAAAAGTCGATGGG
CCGTTAGTCGAGGCTCCGGCCATTATTCGAACTGATCGCGGCATCTATTATCTGTTCTTC
TCTTCGCATTGCTTTACATCGTCGAAGTATAGTGTTAAGTATGCATGGTCAACATCGTTA
AAGGGGCCGTATACCAGGGCTGAGAGGCCCTTGTTCCGGTCTGGGGACTTTGGGCTAAAG
TCTCCTGGCGGTGCGACGGCTTCGGTTGATGGGTCTAGGATTGTTTTTCATGCCTTTTGT
GGGGATTATAGATGTATGTATGCCGCGGCGATGAATATTACGGCTAATTATACGATTTTG
CCGGCGGCGCTTTGA

>XM_001400147.2 Aspergillus niger CBS 513.88 arabinan endo-1,5-alpha-L-arabinosidase C, mRNA
TTCTTTGTTTTTTTTCCCCTCCTCCATTCCCCTTTACCCTCCTCTTCCCTTTTCTTCTTTTCCCCTCTGGAGGATACCCCCTTATCCGCTATGCTTTCGTTTGTCTTGCTTCTTTGTGTAGCGTTGGTCAACGCCTACTCAGACCCGGGTGCATGCTCGGGAACCTGCTGGGCACACGACCCCAATGTCATTCGCCGTGTGTCGGATGGAACCTACTTTCGTTTCTCGACAGGTGGTGGTGTCCATATCTCCTCTGCCAGCGCCATCACTGGTCCCTGGACTGATCTCGGGTATGCACTGCCTAACGGATCCATCGTTACAGTGGGAAATGCCTCCAACCTTTGGGCTCCGGACGTACACTACGTAGATGGCACATACTACATGTACTATGCTAGCTCTACCCTGGGCAGCCGGGATTCCACCATTGGAGTTGCAACCTCCACCACCCTGGAAGCCGACTCCTGGACCGACCACGGCGAGATTGGTGTCACCTCGTCTTCGTCCACCCCTTACAACGCCATTGACCCCAACTGGATCACCATTGGCAGCACCCCCTATCTCCAATTCGGCTCCTACTGGCAAGGTCTCTACCAGGTGGAAATGACCGACTCCCTGAGCGCCAGCAGCAGCACGCCCACTAACTTGGCCTACAACGCATCGGGGAACCATGCTATCGAGGCTTCTTACCTGTACGAGTACGGAGGCTACTACTACCTCACCTTCTCGTCCGGCAAGGCTCAGGGGTACACGACCTCCCTGCCTGCCCAGGGCGATGAGTACCGCATTGTCGTTTGCCGGTCCAAGACTGGAACGGGTAACTTTGTCGACAAGGATGGTGTTTCATGCCTGAACAGCGGCGGAACCACCGTTCTGGCCAGCCACGACTATGTCTACGGCCCTGGTGGACAGGGTATCATCAACACCACCAGCCACGGTATTGTCGTCTACTACCACTATGCCAACAAGAACATTGGCCTGGCTGTCGATGACTACCAGTTCGGCTGGAACACGCTCACCTGGACTGATGGGTGGCCTGTTGTGGCGTGATTTCATTCTTCTTCTTCTTCTATACATACATACATTTTACCGGCGGCCCAGAAATGGGCTCGGTAATATTTGCATGATGATACCTGCGTTCGATTGTATTTAGCTAATTACGTTAATTCGTAATGTTTTATATTGCGATTTAATATATACATATTGAGACATAATTTGTTTGAGAGATTTGTGTAGAATGTCTGTAGAAAATATGAAGCGTCTCGAATTAAGAATGTCAACAACAGGAGTCACTAGAAGGGGCAGTTTAATGATGAAATTACTATCACCCTACTTGATCCTTAATTCCGTCACTCCCCCTTTGACATTCCCCATCCCATCCCCATACAAAAAGCAAAAATGCCACCCCATAAAAAAAAAGAGCCTAGACAACAAACAAACAAACATAACATACCATCCTACGGACCGCACCTAAACTCAATTGTGGAACAATCAGGATTCCTCGGTATTTTGCCCCGAAGAATCTACTACATACCCACATTACTATGCACATTGCCATTGCCACCCAAGCGAAACAAACAACTCACCACATACTGATTAATTAAACGCACAATCCCCAATGCAA

>XM_001399243.2 Aspergillus niger CBS 513.88 arabinan endo-1,5-alpha-L-arabinosidase B, mRNA
ATGACCCCCAACTGGTCCAAACTATGGACCTTCATCGCCAACCCCAAAGACCCCAGTTCTAGCTCCCCATCACCATATACACTCCGTCGCATAATCAAATCCCTCTCCCTTCTCACCGTCTTCTCCATCTTCCTCTACGCCCTCTACATTCACTTCCAGCCCTCCATTATCCCACAAACCCCAGACCTCCCAGACCCAGACCTCCCTCCCTCCCCAGAAGACTCATACAAATCCATCTACGGCTACCCCCCAACCAACCCCACCATCCCTCCCCTGCACATCCACGACCCCAGCATCCTCTACGACCTGCCCACAAACACCTACTACGCCTACGGCTCCGGCCCTCACATCCCCATCCACTCCGCCCCGACCCTCCAAGGACCATGGACCAAAGTCGGCACCGTCCTCGATGCAGATAGTATTCTACCAAAGGGTGATCGCAAAGCCCCATGGGCACCGACCGCCCTCGTCCACGACGGTACCTTCTACGTCTTTTACGCGACGAGCCATAGTGGATGTCGCGATAGCGCTATCGGCGTGGCTACCTCTACTTCTCCGGGCCCTGGGGGATGGGAGGACCACGGGGCGATAGCTATCTCCGGACGAGGGGAGAGGGGGAAGGAGTACCCGTTTGATAGGGCGAATGCGATTGATGTTAGTGTTGTTAAAGGAAAAGGGAGAGGATACATGACCTTCGGCAGTTTCTGGACGGGGATATGGCAAGTCCCGCTGAAGCCTAACCTGCTTCATATGGACAAGCAGGGAGAAGAAGAAAAAAGGGTTAAACATCTCGCCCACGAGCCCGCAGCCATCCACCCACCAACCAAAAAAGCAGATGGATTATGTGGTGATACGACGGGCATGCATCCCATCGAGGGGGCGTTTATATCCTATCATGAGCCGTGGTGGTATTTGTGGTTTAGTTGGGGGAAGTGCTGTCATTTTGATCCGGAGAAATTGCCGAGGGCTGGTCTTGAATACAGCATCCGCGTCGGACGATCAAGTTCTCCTCAAGGTCCATTCGTGGATAAAGAAGGGAAAGATCTAGTGGATGGAGGAGGGGAGATTGTGTATGGGTCGAATGGGGATGTTTATGCGCCTGGGGGACAGGGCGTGTTGAGTGGGGAGGTGGAGGGAGATGTGCTTTATTATCATTATTTGAATATATCTGTGGGGTATGAGTTTAAGGAAGCACGGCTGGGATATAATTATTTGAAGTATGTGGATGGGTGGCCGGTTCCACTGTGA
>XM_001389919.2 Aspergillus niger CBS 513.88 glucan endo-1,6- -glucosidase BGN16.3, mRNA
ATGAGTTGGATGTTGTGGTATACTATAGCTGGCCTAGCATCTCTAGGCCATGCTTCTCTTGAGAAACGCGCCACCGATGCATCTGCGTACTGCTCCGATTCATCGGGTACCTACAAGCTCTCGTCGATCACGGCCCCAGTTCAGGGAAGTGGCAGCCCGGGCTCAGCATCCACCTGGAAGCTAAGTATTGACGATACATCATCCGGTCACAAGCAGGCAATCACTGGCTTCGGGGCTGCTGTGACCGATGCGACCGTCACGTCTTTTAACACCTTGTCGAGCTCTACCCTCCAACAGCTCCTCAATGAGTTGATGACTTCCGATGGAGCTAATTTCTCTCTCATGCGCCATACCATCGGCGCGTCTGATCTGTCCGGTGATCCGGTCTACACCTATGATGACAACGGAAACGAAGCAGACCCCTCACTTTCCGGCTTCGGTCTCGGCGACCGTGGCACTGCTATGGCGGAAATGTTGGCCACGATGAAATCGTTGCAGTCAGACCTGAAGGTGCTTGGGTCCCCGTGGAGTGCGCCCGGATGGATGAAATTGAATGGCGTGATTGATGGCAACGCGACTGATAACAACCTTAACGATGGCTACCTTACCAGTGGTGGCACGGGAAGCACCGGCTACTCCAGTGCTTTCGCGCAGTACTTCGTGAAATACATCCAAGCGTACGAAGCATTAGGAGCTACGGTCGATGCCATTACGATTCAGAATGAGCCGTTAAATAGTCAATCTGGATATCCCACCATGTATGTTTACGACTATGAGTCTGCGCAGCTGATCCAAAAATACATTGGACCGGCTCTTGCCGACGCTGGACTGAGCACGGAAATCTGGGCCTTCGATCACAATACTGATCACGCATCGTACCCTCAGACTGTGCTTGATGAAGCAAGTCAATACGTCGACACGGTAGCCTGGCACTGTTATGCAAAGGATGTTGACTGGACTGTTCTCAGCCAGTTCTACAGCGCCAACCCTGGGGTCAAGCAATATATGACCGAATGCTGGACACCAGCTTCCGGATCATGGCACCAGGCTGCTGATTTTACCATGGGCCCTCTGCAGAACTATGCGTCGGGGGTCATGGCGTGGACTCTGGGCACGAACACTGACGACGGTCCGCATTTGACCAGTGGTGGATGCGACGCCTGTCAGGGCTTGGTCACCATCAACGATGATGACAGCTATACTCTTAATACTGCCTACTACATGATGGCACAGTTTAGCAAGTTCATGCCTCCTGGCGCGACCGTGTTGAACGGAACGGGGAGCTATACCTACTCGGGTGGCGATGGCGTACAGTCGGTGGCGTCCCTGAACCCTGATGGAACGCGCACAGTCATTATTGAGAATATATTCAGTAACGATATTTACTTGACTCTGTCGACGGAGAGCGGAGAGGAGTGGAGTGGAAATATCCCCAGTGAATCGGTGACTACTTGGGTCCTTCCGGCCTTGTCTTGA

>XM_001399115.1 Aspergillus niger CBS 513.88 glycosyl hydrolase, mRNA
ATGTACATCAACCCCATCCTACCGGGATTCAATCCCGACCCCTCTATTATCCGCGTCGGGAATGACTTCTTCCTCGTCACCTCCTCGTTCGAATACTTCCCCGGTGCCCCCATCTACCACAGCACAGACCTTATCCAGTGGAATCTGATCGGCCATGCCCTCACCCGTCCCAGCCAACTCCAGATTCAAACCCCCGAGCCTGGTGGTGGCGTGTGGGCCACCACTCTGCGGTATCACGATGGCGTATTCTACATCCTCGCCGCAAGCTTCCAGCGTTACCGGCCACAGGAAGATGACCGCGTCTGGCCGCGTGGTTTCTATGTGAAGACCAGAAATATATGGGACTCGAGTTCGTGGTCGGACCCGGTGTATTTTGACCAAGTTGGCTTTGATCAAGATCTATTTTGGGACGACGATGGCACCGTCTATCTCTCAAGCACGTATCGCAAGCTTCACCGAACTCCCGGTGTGAACCTCAAGGATTTCGCCGTTCACATTGTCACTGTCGACCTGGAGACAGGTGCCTCCACTTCGGAGCCCAAGTTAATCAGAGAGTCCACCTCCGGTGTCGCGGAGGGGTCGCATATTTTCAAACGCGGCAAATACTACTATCTTTTCACCGCGGAGGGCGGCACGGAGAGCGGTCATTGCGAGTGGGTCAGTCGCAGTACTACTGGCCCGTTCGGGCCATGGGAGATTGGGCCCAATAATCCTCTCTGGAGGAATGGAGTGGATGATGAAGTTCAGAACACTGGGCATGCAGACCTGGTGGAAGATGCGCAGGGGCAGTGGTGGGCTGTTTTGTTAGGTGTGCGTCCTGTCCGGAAGGATGGTCAGTGGGAGGAATCGGTTTTTGGTCGCGAAACATTTCTAGCTCCTGTGAACTGGGAGAACGACTGGCCCATTGTCAACGGTGGCCAAAAGATCCCCCTCCACTCTGAGAGCCCCCACCTCTATGAGCTCTCAGTTCCAGTCACCTGGCGAGATGACTTTTCGGACAGTGAGCTCCAGCTAGGATGGTACCGCAAGAATACTCCAGTTGTGCGTGACTACTCTCTCACCGAGCGACCAGGCCACCTTCGCCTCTACGGTGGTCCATATAATCTGTCCGTCCCGGCGTGTCCAACTCTATTTCTTCGTAAACAGACCCATCGGTTCTGTACATGGGAGACACTGCTGTCTTTCCATCCGAGGAACCAGGCATGCGAGGCTGGTGCAGTCGTCTGGTGGACGTACTTCACCTACAGCAGCATCGGGATTTGTCGAGCGAACGATGGGCAGCGTGTTGTTCGGCTACGTTCTCCGGGAGGTGAGATAGTCGACTTCGGCCTACAGAATTCGCAGTCGGATGTCCGTTTGGTGATCGAGTGTGGGGATCGCTACCGGTTCGGATTTCAGGAGATCAGAGAGGATGAGTCGAATGAAGAGCCGATACGGTGGATCGGCGAGGCGGAGAATCGCAGTATGACAGCAGCACCGCCGGTAGGAGTGGCTTTTACGGGGATGATGGTGGGATTGTACGCGTTTGGAGAGCGGCAACGGTGTCTAGATCCAGCTGACTTTCATTACGCGTTATGGAGATAG
```

## Output

cupp-predict outputs two files by default. These are a tab separated file (TSV) containing each annotation associated with an input sequences, and an annotated FASTA file containing each sequence with annotations.

The following also applies to both files

- If a CUPP group contains more than one EC number or SubFamily in either the BEst or All columns, then these are separated by a "-".
- The "&" symbol indicates that the EC or SubFamilies have been experimentally shown to exist in the same protein. (A CUPP group contains a number of CUPP tides from a number of proteins that cluster together based on similarity of CUPP tides).
- If no EC is known for a CUPP Group then "Unk" will be given in the EC_Best and nothing in EC_all
- If no SubFamily is known then the SubFamily_Best and SubFamily_All are left blank
- If either the CUPP Group's significance score does not pass the "--cuppconfidence" or "--subfamilyconfidence" thresholds, then no EC or SubFamily will be given but replace by "!". In addition, the CUPP_Group will be prefixed also by "!".

### TSV

The primary output of CUPP predication is a tsv file containing the relevant scoring, position and description for each annotation.

The layout is as follows:

- A number of header lines prefixed by "#" that contain version information for the CUPP CLI and the CUPP library versions
- A header line also prefixed by an "#"
- The annotation on individual lines with the following columns. (Note a single entry sequence can have multiple annotations and thus multiple lines with the same accession)
  - Accession - The header for this FASTA entry in the input file
  - Family - The predicted family based on the CUPP Group annotation
  - CUPP_Group - The predicted CUPP Group that this annotation is associated with
  - Significance - The significance score of the annotation. This is used for filtering and also the confidence of the EC and SubFamily annotation
  - EC_Best - The best probable EC number that is accosted with the CUPP Group
  - SubFamily_Best - The best probable SubFamily number that is accosted with the CUPP Group
  - Start - Start position of the annotated domain
  - End - Start end of the annotated domain
  - Coverage - Start number of positions within the domain that are covered by CUPP tides
  - Protein_Length - The full length of the original FASTA entry sequence or ORF
  - EC_Full - A list of other possible EC numbers that are accosted with the CUPP Group
  - SubFamily_Full - A list of other possible SubFamilies numbers that are accosted with the CUPP Group

```
# Conserved Unique Peptide Patterns (cupp) Prediction!
# Version: 3.1.1
# Library file version: 2020.2
# Meta file version: 2020.2
#Accession	Family	CUPP_Group	Significance	EC_Best	SubFamily_Best	Start	End	Coverage	Protein_Length	EC_Full	SubFamily_Full
EAA29705.1 hypothetical protein [Neurospora crassa]	GH30	GH30:21.1	64.451	GH30:3.2.1.164	GH30_5	37	433	197	488	GH30:3.2.1.164	GH30_5
BAO73305.1 LcGUS30 [Lactobacillus brevis subsp. coagulans]	GH30	GH30:30.1	272.258	GH30:3.2.1.31	GH30_9	42	474	368	510	GH30:3.2.1.31	GH30_9
BAE55732.1 unnamed protein product [Aspergillus oryzae RIB40]	GH43	GH43:1.1	198.504	GH43:3.2.1.37	GH43_1	8	317	173	332	GH43:3.2.1.37&3.2.1.55	GH43_1
AAP64786.1 Sequence 2 from patent US 6555335	GH30	GH30:26.1	18.614	GH30:3.2.1.*	GH30_7	86	380	61	465	GH30:3.2.1.*&3.2.1.8	GH30_7
AAU40201.1 arabinan endo-1,5-alpha-L-arabinosidase AbnA [Bacillus licheniformis DSM 13 = ATCC 14580]	GH43	GH43:19.1	293.073	GH43:3.2.1.99	GH43_5	49	325	217	328	GH43:3.2.1.99	GH43_5
```

### FASTA

The secondary output is a FASTA file of the annotated sequences (or all sequences if the "--keep" flag is given). These headers of the fasta entries are formatted to contain the information about the annotated domains that the entry is predicted to contain. Likewise, the domain is highlighted in the sequence through capitalisation.

If multiple domains are found in an entry FASTA, then there are listed in order of position, separated by "+".

The header is formatted as such:

```
>HEADER IN ORIGINAL FASTA|CUPP_GROUP(START..END)|EC_BEST|SUBFAMILY_BEST
seq...

>HEADER IN ORIGINAL FASTA|CUPP_GROUP_1(START..END)+CUPP_GROUP_2(START..END)|EC_BEST_1+EC_BEST_2|SUBFAMILY_BEST_1+SUBFAMILY_BEST_2
seq...
```

Full example:

```
>EAA29705.1 hypothetical protein [Neurospora crassa]|GH30:21.1(37..433)|GH30:3.2.1.164|GH30_5
mvpslqtvghilvtalaiipfafadltttidpssnwGTWEGWGTSLAWWAATFGTRDDLA
DIFFTTQTTNLNGVSLPGLGFNIARYNAGASSWNSINGTSMVVSPKMIPSRQIEGFWIDW
ASTSPSSASWNWSADSAQRAMLSKARSRGANILELFSNSPMWWMCNNHNPSGSDDGSSDN
LQSWNYQSHAVYMATVAKYFADHWSTTFTSVDPFNEPSANWWNGKTGTQEGCHFDVSTQA
TVLNYLRAELDSRGLSSTVISASDESYYDQAAKTLQNLGSTALGKIARVNVHGYQYGSGR
RDSVRSLATQYSKKLWQSEYGEGDATGRDLASNLLLDLRWLQPTAWVYWQALDGGGWGLV
DADNDKKTIGQPTQKYFVLAQFARHVKPGMRILDGGSDYATAAYDAATKKLVIVAVNWGS
AQYLNFDLSRFKTagvngavvkrwatqigastgkryvaasdtyirgtkfwsyfekemvqt
fevegvvl

>BAO73305.1 LcGUS30 [Lactobacillus brevis subsp. coagulans]|GH30:30.1(42..474)|GH30:3.2.1.31|GH30_9
mkkfqhefwtatsgdltqrrtplatpdyqstsaaavkiiidPADRHQPWLGGGAAITDSA
AYLLWSVMSAEQRRALLTELFDPDQGGFSSVRVPLGSCDFQSQDFYTYDDVPYGEHDQKL
EQFSIGTGQPGAPDATKDLKHIVPVLQEILAINPAVKVIASPWSAPAWMKNTGHLTHGGH
LRFGEFTGNGYTEENRFEYIYAQYFIRYIEAYQKLGIPIYGLTIQNEPSNAAHWPAMIWT
VPQLADFGYRYLRPALNHSFPDTKLYLLDDSFHALTKPITAEVTPEQAAAFDGLAVHTYS
GPYDNLYHANRAYPNWSTIMTERRCMMTDTPEEAAHIMFGIIGNWLVHNGLSMITLWNLA
LDERGLPNAADSTGREGVVTIDHTTGKVQRNLEYFMLRNFGQDVSVGATVIGSTNYTRDG
YTGGLGSVAFLGTAGDIAAHLYNPTAQPIQAAVTINGNGANWQLVTVPPYGTVTlhksda
plnttnvpvddefplnptpanhsdvapgqh

>BAE55732.1 unnamed protein product [Aspergillus oryzae RIB40]|GH43:1.1(8..317)|GH43:3.2.1.37|GH43_1
msnsnppLITHIYTADPSAHVFNNKVYIYPSHDRETDIQFNDNGDQYDMVDYHVFSMDSI
SGPVTDHGVVLTLSDVPWASKQLWAPDAATKDGKYYLYLPARDHDGIFRIGVAVAEKPEG
PFTAQKSFIAGSYSIDPASFVDDDGKAYLYFGGIWGGQLQCWSKDAGAAGEWVFDASKTG
PQEPSGENELALHPRVAQLTDDMLGFETPVQELVILDQDGVSPLKADDHERRFFEASWMH
KYNGTYYFSYSTGDTHYIVYATGESPFGPFRYRGRILEPVLGWTTHHSIVEFGGRWWIFY
HDCELSKGVSHLRSVKVreigyddqgdiylvd

>AAP64786.1 Sequence 2 from patent US 6555335|GH30:26.1(86..380)|GH30:3.2.1.*|GH30_7
mkssisvvlallghsaawsyatksqyranikinarqtyqtmigggcsgafgiacqqfgss
glspenqqkvtqilfdenigglsivRNDIGSSPGTTILPTCPATPQDKFDYVWDGSDNCQ
FNLTKTALKYNPNLYVYADAWSAPGCMKTVGTENLGGQICGVRGTDCKHDWRQAYADYLV
QYVRFYKEEGIDISLLGAWNEPDFNPFTYESMLSDGYQAKDFLEVLYPTLKKAFPKVDVS
CCDATGARQERNILYELQQAGGERYFDIATWHNYQSNPERPFNAGGKPNIQTEWADGTGP
WNSTWDYSGQLAEGLQWALYMHNAFVNSDTSGYTHWWCAQNTNGDNALIRLDRDSYEVSA
RLWAFAQYFRFARPGSVRIGatsdvenvyvtayvnkngtvaipvinaahfpydltidleg
ikkrklseyltdnshnvtlqsrykvsgsslkvtvepramktfwle

>AAU40201.1 arabinan endo-1,5-alpha-L-arabinosidase AbnA [Bacillus licheniformis DSM 13 = ATCC 14580]|GH43:19.1(49..325)|GH43:3.2.1.99|GH43_5
mlktskferrttvaktiisgfilffllifsaaeptsaafwdtkgdniiHDPSMIKEGNTW
YTFGTGIGNGIRVIKSTDGKTWSAAPSIFATPLSWWKKYVPNHEENQWAPDISYYNGRYW
LYYAVSAFGSNTSAIGLASTDRISSGNWRDDGLVIRTTTSNNYNAIDPELVIDKEGNPWL
SFGSFWSGIKLTKLDKKTMKPTGKVYSIASRPNNGGAVEAPSITYKNGYYYLFVSFDKCC
QGVNSTYKIAYGRSTKITGPYYDKSGKKMTEGGGTILDSGNDQWKGPGGADIVNGNIIVR
HGYDALDNGAPKLLINDLNWDSKGWpky
```

### Definitions and algorithm

CUPP prediction works to assign a CUPP Group to a certain domain of a protein sequence. This CUPP Group is contained in a predefined library included alongside the cupp-rust cli. The CUPP Group has a number of features associated with it that are related to CAZy annotation. These include Family, Subfamily and enzyme class (EC).

An iterator generates small sequential peptide fragments (CUPP tides) along the length of the protein sequence in sliding window manner, as well as replacing 2 position with the wildcard character "X". These tides are subsequently check for existence in the CUPP library and if contained within, that CUPP group is given a score based on the score of the CUPP tide. After all tides are processed, if the summed score for that CUPP Group is above a threshold, then a domain with a start, end and coverage is generated for this hit group. These domain ranges are then filtered for overlap and significance and any that pass this filtering stage are written to the output files.

More information can be found at the citations below.

### Todo

Some features are still under development:

- CUPP clustering
- Accepting gzip files as input

### Contributors

Cameron H. <cjahu@dtu.dk>

Kristian B. <kbaka@dtu.dk>

### License

The following license apply to this project through its dependencies:

- MIT License
- [Apache License 2.0](https://www.apache.org/licenses/LICENSE-2.0)
- [Mozilla Public License 2.0](https://www.mozilla.org/en-US/MPL/2.0/)

Or view the license files included.

Parts of CUPP predict are based on the FASTA reading function in the rust-bio library which is available at [rust-bio](https://github.com/rust-bio/rust-bio).

The relevant reference can be found here:

Köster, J. (2015). Rust-Bio: a fast and safe bioinformatics library. Bioinformatics, 32(3), 444–446. https://doi.org/10.1093/bioinformatics/btv573

### Citation

When using the CUPP programs or results, please cite:

Barrett, K., Lange, L. Peptide-based functional annotation of carbohydrate-active enzymes by conserved unique peptide patterns (CUPP). Biotechnol Biofuels 12, 102 (2019). https://doi.org/10.1186/s13068-019-1436-5

The citation for the webserver version is:

Kristian Barrett, Cameron J Hunt, Lene Lange, Anne S Meyer, Conserved unique peptide patterns (CUPP) online platform: peptide-based functional annotation of carbohydrate active enzymes, Nucleic Acids Research, Volume 48, Issue W1, 02 July 2020, Pages W110–W115, https://doi.org/10.1093/nar/gkaa375
